### PR TITLE
AABB improvements

### PR DESCRIPTION
**Pull request by eviltak**
## Goal
Adds documentation to the undocumented methods in the `AABB` class, and performs some refactoring.

## Documents
- `createMinMax`
- `createCenterExtent`
- `createEmpty`
- `createEncompassing` (vertices variant)
- `getCenter`
- `getMin`
- `getMax`
- `transform` (both variants)
- `getFirstHitPlane`
- `intersectRectangle`
- `minX`, `Y`, `Z`
- `maxX`, `Y`, `Z`

## Removes
- `AABB.getFirstHitPlane`

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 1]
```diff
@@ -58,6 +72,11 @@ public static AABB createCenterExtent(Vector3f center, Vector3f extent) {
         return new AABB(min, max);
     }
 
+    /**
+     * Creates an empty AABB that does not encompass any points.
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: ~encompass~ contain

Comment (mjuvekar7) [reply engine/src/main/java/org/terasology/math/AABB.java]: Do this everywhere. Sounds much more natural

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 2]
```diff
@@ -88,6 +107,12 @@ public static AABB createEncompassing(Iterable<AABB> others) {
         return new AABB(min, max);
     }
 
+    /**
+     * Creates a new AABB that encapsulates the vertices as represented by a {@link TFloatList}.
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: ~encapsulates~ contains

Comment (mjuvekar7) [reply engine/src/main/java/org/terasology/math/AABB.java]: Also ~the~ and ~as~

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 3]
```diff
@@ -117,17 +142,26 @@ public Vector3f getExtents() {
         return dimensions;
     }
 
+    /**
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: Description!

Comment (mjuvekar7) [reply engine/src/main/java/org/terasology/math/AABB.java]: Even the most trivial methods must have a description

Comment (mjuvekar7) [reply engine/src/main/java/org/terasology/math/AABB.java]: Same for all other get functions

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 4]
```diff
@@ -145,11 +179,25 @@ public AABB move(Vector3f offset) {
         return new AABB(newMin, newMax);
     }
 
+    /**
+     * Transform this AABB into a new AABB with the given rotation, offset and scale.
+     *
+     * @param rotation The rotation from the current AABB to the new AABB.
+     * @param offset The offset between the current AABB and the new AABB.
+     * @param scale The scale of the new AABB with respect to the old AABB.
+     * @return the new transformed AABB.
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: Capital T

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 5]
```diff
@@ -345,26 +404,44 @@ public Vector3f centerPointForNormal(Vector3f normal) {
         return new Vector3f();
     }
 
+    /**
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: Description. Same for all other min functions

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 6]
```diff
@@ -88,6 +107,12 @@ public static AABB createEncompassing(Iterable<AABB> others) {
         return new AABB(min, max);
     }
 
+    /**
+     * Creates a new AABB that contain the vertices as represented by a {@link TFloatList}.
```
Comment (mjuvekar7) [inline engine/src/main/java/org/terasology/math/AABB.java]: contain**s** (small grammatical mistake)

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 7]
```diff
@@ -46,10 +46,24 @@ private AABB(Vector3f min, Vector3f max) {
         this.max = max;
     }
 
+    /**
+     * Creates a new AABB from the given minimum and maximum points.
+     *
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: Can you add an example like if min or max are inclusive or exclusive?

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 8]
```diff
@@ -46,10 +46,24 @@ private AABB(Vector3f min, Vector3f max) {
...
+     * @param max The maximum of the AABB.
+     * @return The created AABB.
+     */
     public static AABB createMinMax(Vector3f min, Vector3f max) {
         return new AABB(new Vector3f(min), new Vector3f(max));
     }
 
+    /**
+     * Creates a new AABB with the given center and extents.
+     *
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: Can you provide an example? Like creating an AABB around 0,0,0 with extent like 1,2,3. What would be the result?. At least that is something I would like to know when reading the documentation.
The current sentence adds no more information than the method and parameter names which already exist.

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 9]
```diff
@@ -88,6 +107,12 @@ public static AABB createEncompassing(Iterable<AABB> others) {
         return new AABB(min, max);
     }
 
+    /**
+     * Creates a new AABB that contains the vertices as represented by a {@link TFloatList}.
+     *
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: Important information for a caller would be the format for vertices and what would be the result. An example would be nice. E.g. this method spans a bounding box which contains all vertices.

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 10]
```diff
@@ -88,6 +107,12 @@ public static AABB createEncompassing(Iterable<AABB> others) {
...
+     * @param vertices The vertices to encompass.
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: Vertices like x,y,z,x2,y2,z2, .... Describe the expected format here.

File: engine/src/main/java/org/terasology/math/AABB.java (hunk 3 above)
Comment (oniatus) [reply engine/src/main/java/org/terasology/math/AABB.java]: I would like to see javadoc where it adds more information which is not visible in the API, for example what elements are expected to be in a passed list, if a method can handle null, if a method has side-effects, all these kind of things.
Javadoc which repeats the method name like
```
/**
* Getter for foo.
* @return the foo.
**/
public Foo getFoo();
```
are just 4 lines of wasted space and 10 seconds wasted of developer time, they add zero new information.

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 11]
```diff
@@ -117,17 +142,32 @@ public Vector3f getExtents() {
...
         dimensions.add(min);
         dimensions.scale(0.5f);
         return dimensions;
     }
 
+    /**
+     * Returns the minimum of the AABB.
+     * 
+     * @return The minimum of the AABB.
+     */
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: I think we can remove these, there is 0 logic here to explain.

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 12]
```diff
@@ -117,17 +142,32 @@ public Vector3f getExtents() {
...
     public Vector3f getMin() {
         return new Vector3f(min);
     }
 
+    /**
+     * Returns the maximum of the AABB.
+     * 
+     * @return The maximum of the AABB.
+     */
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: Same for this one, get rid of it :)

File: engine/src/main/java/org/terasology/math/AABB.java [hunk 13]
```diff
@@ -224,6 +278,17 @@ public Vector3f closestPointOnAABBToPoint(Vector3f p) {
         return r;
     }
 
+    /**
+     * Sweeps an AABB on to the AABB in the given direction and returns the normal of the first plane of intersection.
+     *
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/math/AABB.java]: @flo and I double-checked this method, it includes some weird behavior, e.g. if all flags are set the normal will be overridden and only the value for Z will be returned.
Besides that there is no single use of the method in the engine and also none in any module.

Please remove the entire method ;)

File: engine/src/main/java/org/terasology/math/AABB.java (hunk 5 above)
Comment (oniatus) [reply engine/src/main/java/org/terasology/math/AABB.java]: We should keep either the description or the return statement (my favorite would be the return statement and an empty method description).
Redundancy is always a bad thing ;) keep it DRY.

Same for all further min.. and max.. methods.

File: engine/src/main/java/org/terasology/math/AABB.java (hunk 3 above)
Comment (eviltak) [reply engine/src/main/java/org/terasology/math/AABB.java]: @oniatus Excellent point. Should I remove the docs in all such cases?

File: engine/src/main/java/org/terasology/math/AABB.java (hunk 5 above)
Comment (eviltak) [reply engine/src/main/java/org/terasology/math/AABB.java]: Definitely have to agree. Will keep that in mind.

File: engine/src/main/java/org/terasology/math/AABB.java (hunk 7 above)
Comment (eviltak) [reply engine/src/main/java/org/terasology/math/AABB.java]: Min and max should be inclusive. However, I do not understand what you mean by adding an example for this. What kind of example?

**Comment by Cervator** [comment]
Bump - @oniatus & @eviltak you might both have comments to respond to here :-)

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by jellysnake** [comment]
I've just had a read over this, apart from the comments left by oniatus, it looks good to me :)

If we can get those few things tweak I think we should merge this

**Comment by mjuvekar7** [comment]
Bump @oniatus @eviltak. Maybe it's time to take care of this? :)

**Comment by Cervator** [comment]
Merging this with a few added trivial tweaks - can usually always improve docs further, but this is a step forward and we need to chew through some PRs already :-)

**Pull request merged**